internal/jobobject/jobobject.go

```diff
@@ -0,0 +1,414 @@
+// This call will return immediately with error `ErrNotRegistered` if the job was not registered
+// to receive notifications during `Create`. Internally, messages will be queued and there
+// is no worry of messages being dropped.
+func (job *JobObject) PollNotification() (interface{}, error) {
+	if job.mq == nil {
+		return nil, ErrNotRegistered
+	}
+	return job.mq.readOrWait()
+}
+
+// Close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *JobObject) Close() error {
+	job.handleLock.Lock()
+	defer job.handleLock.Unlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	if err := windows.Close(job.handle); err != nil {
+		return err
+	}
+
+	if job.mq != nil {
+		job.mq.close()
+	}
+	// Handles now invalid so if the map entry to receive notifications for this job still
+	// exists remove it so we can stop receiving notifications.
+	if _, ok := jobMap[uintptr(job.handle)]; ok {
+		delete(jobMap, uintptr(job.handle))
+	}
+
+	job.handle = 0
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *JobObject) Assign(pid uint32) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+
+	if pid == 0 {
+		return errors.New("invalid pid: 0")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, pid)
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.handle, hProc)
+}
+
+// Terminate terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *JobObject) Terminate(exitCode uint32) error {
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	return windows.TerminateJobObject(job.handle, exitCode)
+}
+
+// Shutdown gracefully shuts down all the processes in the job object. Context timeout
+// will be used as timer for graceful shutdown. If all of the processes in the job have not
+// exited by the timeout, the job will be terminated.
+func (job *JobObject) Shutdown(ctx context.Context) error {
+	job.handleLock.RLock()
+	if job.handle == 0 {
+		return ErrAlreadyClosed
+	}
+	job.handleLock.RUnlock()
+
+	pids, err := job.Pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Errorf("failed to send ctrl-break to process in job: %s", err)
+			return job.Terminate(1)
+		}
+	}
+
+	// Check if the processes have exited, it may take some time for some processes to clean up
+	// so keep checking until we've either hit the timeout set on the context or the processes
+	// clean up successfully.
+	for {
+		time.Sleep(time.Millisecond * 10)
```

Comment:
Is there any big side effect from not having this line in?

Reply:
Just basically giving time for anything to progress (processes to actually clean up) instead of the second the pids call is done and we've checked the pid count to just call it right again.

Reply:
I understand the usefulness of it, but at the same time, this creates an artificial bottleneck here (even if it's this small lol) where the minimum amount of time in which the processes could have exited would be 10 milliseconds. Maybe move this to the end of the for loop somehow?

Reply:
Nothing stopping it from being at the end here. Can move

Reply:
Jobs send a notification when all the processes exit. Can't we just use that instead of polling for pids to determine when everything has exited?

Reply:
What if we specified we don't want notifications?

Reply:
Well if notifications are required for core job object functionality (like `Shutdown`), then we presumably wouldn't want to give the caller an option to disable them.

However, in this case I'm thinking we should just remove `Shutdown`. There's nothing being done that couldn't be implemented by a client of the `jobobject` package, and all it's really doing is sending a signal to some processes then waiting, which isn't any functionality related to job objects.

Reply:
Haha the life of how notifications have evolved in this package has been a storied one 😄. I can remove `Shutdown` and if you feel we shouldn't have a way to turn off notifications then I can also remove this although a little more hesitant here. If a client didn't really care about the messages it does seem wasteful though, we have messages sitting idle in the kernel and a local copy on our end in the queue as well.

Reply:
I don't mind if we leave in the notifications bool on Create, but let's remove Shutdown in that case.

Reply:
Sweet, on it. Thanks Kevin

Reply:
Done